isic-archive-tasks/isic_archive_tasks/zip.py

```diff
@@ -0,0 +1,126 @@
+    """
+    Ingest images from a ZIP file into a dataset.
+
+    The images are extracted to a "Pre-review" folder within the dataset folder.
+    """
+    from girder.plugins.isic_archive.models.batch import Batch
+    from girder.plugins.isic_archive.models.dataset import Dataset
+    from girder.plugins.isic_archive.models.image import Image
+    from girder.plugins.isic_archive.upload import ZipFileOpener
+    from girder.plugins.isic_archive.upload import TempDir
+
+    dataset = Dataset().load(datasetId, force=True)
+    user = User().load(userId, force=True)
+    batch = Batch().load(batchId)
+
+    if batch['ingestStatus'] in ['extracted', 'notified']:
+        return
+
+    prereviewFolder = Folder().createFolder(
+        parent=Dataset().imagesFolder(dataset),
+        name='Pre-review',
+        parentType='folder',
+        creator=user,
+        public=False,
+        reuseExisting=True)
+
+    # Get upload information stored on batch
+    s3BucketName = batch.get('s3BucketName')
+    s3ObjectKey = batch.get('s3ObjectKey')
+    if not all([s3BucketName, s3ObjectKey]):
+        raise GirderException('Error retrieving upload information.')
+
+    # Move file from S3 to the assetstore, attached to the dataset
+    s3 = boto3.client('s3', endpoint_url='http://localhost:4572')
+    with TempDir() as tempDir:
+        # Download file from S3 as upload user
+        fileName = os.path.join(tempDir, str(batch['_id']) + '.zip')
+        s3.download_file(
+            Bucket=s3BucketName,
+            Key=s3ObjectKey,
+            Filename=fileName
+        )
+
+        # Attach file to dataset
+        # fileSize = os.path.getsize(fileName)
+        # with open(fileName, 'rb') as fileStream:
+        #     zipFile = Upload().uploadFromFile(
+        #         obj=fileStream,
+        #         size=fileSize,
+        #         name=fileName,
+        #         parentType='dataset',
+        #         parent=dataset,
+        #         attachParent=True,
+        #         user=user,
+        #         mimeType='application/zip'
+        #     )
+        # # TODO: remove this once a bug in upstream Girder is fixed
+        # zipFile['attachedToType'] = ['dataset', 'isic_archive']
+        # set item id to null
+        # zipFile = File().save(zipFile)
+        # propagatesizechange on the item -filesize
+
+        # Delete file from S3 as upload user
+        # s3.delete_object(
+        #     Bucket=s3BucketName,
+        #     Key=s3ObjectKey
+        # )
+
+        # TODO: Could update bucket policy to effectively revoke temporary security
+        # credentials
+
+        # Remove upload information from batch
+        del batch['s3BucketName']
+        del batch['s3ObjectKey']
+        batch['ingestStatus'] = 'extracting'
+        batch = Batch().save(batch)
+
+        # Process zip file
+        with ZipFileOpener(fileName) as (fileList, fileCount):
+            for originalFilePath, originalFileRelpath in fileList:
```

Comment:
We need to add a policy where duplicate **basenames** of files in a ZIP are not allowed (rejecting the ZIP). Also, if any basename of a file in a ZIP matches the `privateMeta.originalFilename` of an image already saved in the dataset, the ZIP should be rejected.

Reply:
I'm going to tackle this in a follow up PR.